# Challenge 1: multiple of 3 and 5

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000

In [ ]:
def get_multiples(base, limit):
    multiples = set()
    
    term = 1
    while True:
        result = base * term
        if result >= limit:
            return multiples
        else:
            multiples.add(result)
        term += 1

In [ ]:
%%time
limit = 10
sum(get_multiples(3, limit).union(get_multiples(5, limit)))

In [ ]:
%%time
limit = 1_000
sum(get_multiples(3, limit).union(get_multiples(5, limit)))

# Challenge 2: Even Fibonacci numbers

Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.


In [ ]:
generated_number_count = 11

In [ ]:
def fibonacci_generator():
    a = 0
    b = 1
    while True:
        yield a
        current_a = a
        a = b
        b = b + current_a

In [6]:
number_count = 0
fib_number_list = list()
for fib_number in fibonacci_generator():
    fib_number_list.append(fib_number)
    number_count += 1
    if number_count > generated_number_count:
        break
print(fib_number_list)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]


In [7]:
def even_fibonacci_generator():
    a = 2
    b = 0
    while True:
        yield a
        current_a = a
        a = 4 * current_a + b
        b =  current_a

In [8]:
number_count = 0
fib_number_list = list()
for fib_number in even_fibonacci_generator():
    fib_number_list.append(fib_number)
    number_count += 1
    if number_count > generated_number_count:
        break
print(fib_number_list)

[2, 8, 34, 144, 610, 2584, 10946, 46368, 196418, 832040, 3524578, 14930352]


In [9]:
%%time 
even_sum = 0
for number in even_fibonacci_generator():
    if number > 4_000_000:
        break
    even_sum += number
print(even_sum)

4613732
CPU times: user 166 µs, sys: 25 µs, total: 191 µs
Wall time: 165 µs


# Challenge 3: Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [10]:
import math

def get_prime_factors(n):
    prime_factors = list()
    primes = list()
    search_limit = int(math.sqrt(n)) + 1
    
    remainder = n
    
    for number in range(2, search_limit):
        if remainder % number == 0:
            multiple_of_existing_prime = False
            for prime in primes:
                if number % prime == 0:
                    multiple_of_existing_prime = True
                    break
            if not multiple_of_existing_prime:
                prime_factors.append(number)
                primes.append(number)
                remainder = remainder / number
    return prime_factors

In [11]:
%%time
get_prime_factors(13195)

CPU times: user 58 µs, sys: 0 ns, total: 58 µs
Wall time: 63.9 µs


[5, 7, 13, 29]

In [12]:
%%time
get_prime_factors(600851475143)

CPU times: user 56.3 ms, sys: 141 µs, total: 56.4 ms
Wall time: 55.6 ms


[71, 839, 1471, 6857]

# Challenge 4: Largest palindrome product

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

## Toughts
First 30 palindromic numbers: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 22, 33, 44, 55, 66, 77, 88, 99, 101, 111, 121, 131, 141, 151, 161, 171, 181, 191, 202

Max value with 1 digit number: 9x9 = 81 (2 digit number) but there is no palindromic number that can be found with 1 digit number multiplication

Max value with 2 digit number: 99x99 = 9801 (4 digit numbers)

Max value with 3 digit numbers: 999x999 = 998001 (6 digit numbers)

In [13]:
def is_palindromic_number(number):
    number_list = [int(i) for i in str(number)]
    reverse_number_list = number_list[::-1]
    if number_list == reverse_number_list:
        return True
    else:
        return False

In [14]:
%%time
print(f'9009 ? {is_palindromic_number(9009)}')
print(f'1 ? {is_palindromic_number(1)}')
print(f'81 ? {is_palindromic_number(81)}')

9009 ? True
1 ? True
81 ? False
CPU times: user 182 µs, sys: 22 µs, total: 204 µs
Wall time: 141 µs


In [15]:
from functools import reduce
import math

def get_factors(n, digit_count):
    if n == 0:
        return 0
    step = 2 if n % 2 else 1
    
    for factor in set(reduce(list.__add__, ([i, n// i] for i in range(1, int(math.sqrt(n))+1, step) if n % i == 0))):
        if len(str(factor)) == digit_count:
            other_factor = int(n / factor)
            if len(str(other_factor)) == digit_count:
                return (factor, other_factor)
    return None

In [16]:
%%time
get_factors(9009, 2)

CPU times: user 18 µs, sys: 2 µs, total: 20 µs
Wall time: 21.9 µs


(91, 99)

In [17]:
def get_palindromic_number(min_digit_number, max_digit_number, max_digit_count):
    number = max_digit_number ** 2
    while number >= 0:
        if is_palindromic_number(number):
            factors = get_factors(number, max_digit_count)
            if factors:
                return number, factors
        number -= 1
    return None

In [18]:
%%time
get_palindromic_number(0, 9, 1)

CPU times: user 104 µs, sys: 0 ns, total: 104 µs
Wall time: 106 µs


(9, (1, 9))

In [19]:
%%time
get_palindromic_number(10, 99, 2)

CPU times: user 781 µs, sys: 97 µs, total: 878 µs
Wall time: 880 µs


(9009, (91, 99))

In [20]:
%%time
get_palindromic_number(100, 999, 3)

CPU times: user 95.6 ms, sys: 3.44 ms, total: 99.1 ms
Wall time: 98.2 ms


(906609, (993, 913))

# Challenge 5: Smallest multiple


2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
